### PR TITLE
Feat: DO-2935 node grouping in fcose layout dara UI

### PR DESCRIPTION
**Pull request by patricia-causalens**
## Motivation and Context

Ability to add groupings which are collapsible in CausalGraphEditor component

## Implementation Description

This PR implements the collapse and expand logic as well as the engine drawing of group nodes. As a start Fcose was adapted to support this feature. 

**Fcose Layout:**

In Fcose layout we take advantage of compound nodes. These are node parents (which here we call group) and all its children are attracted to their position. Meaning we can have the positions of the nodes calculated separated into their own groups. Also added types for layouts that supports groups similarly to how we did it for tiers.

**On the PIXI objects side:**
- I have updated node to have two different textures, the normal circle texture, and a square texture to support group nodes
- In edge calculations the arctangent was defined upside down. I have updated so that it is defined the normal way, and hence updated the corresponding edge calculations. Have also updated to account for square nodes so that edges are still placed on its surface.
- Added a new symbol to edge, the number of collapsed edges, this way we can differentiate between normal edges and group edges more easily in mixed graphs where group nodes and normal nodes may be present.
- Added a new object group container, which is just the container which wraps the nodes when in the expanded state of groups.

**On the Engine side:**
- Added new layers for group containers, including mouse events to be able to capture and display tooltip
- Added a method for collapseAll and expandAll edges
- Disabled editable interaction on group nodes and group edges. This is so the user can't delete a lot of hidden nodes and edges by interacting with collapsed state.

**On the Editor side:**
- Added a new switching button to allow user to switch between collapsed and expanded states
- Added a warning for when users try to add an edge among group nodes. Since it is not clear what should happen to these when the nodes expand this has been forbidden.
- Updated search so that if view is collapsed and we search for a node, view is expanded to show you it

**Other Improvements:**
- Our storybook for graph was getting out of hand with a lot of stories all defined in one massive file. I have broken these down into relevant categories, so that it now has a neater folder structure.
- Fixed an issue with Legends where node and edge threw a warning if they had the same label.

## Any new dependencies Introduced

None

## How Has This Been Tested?

Locally on storybook

## PR Checklist:

- [x] I have implemented all requirements? (see JIRA, project documentation).
- [x] I am not affecting someone else's work, If I am, they are included as a reviewer.
- [ ] I have added relevant tests (unit, integration or regression).
- [x] I have added comments to all the bits that are hard to follow.
- [ ] I have added/updated Documentation.
- [x] I have updated the appropriate changelog with a line for my changes.

## Screenshots (if appropriate):

collapse

### CODE REVIEW

File: packages/ui-causal-graph-editor/src/shared/rendering/utils.tsx [hunk 1]
```diff
@@ -79,3 +79,20 @@ export function getZoomState(scale: number, zoomThresholds?: ZoomThresholds): Zo
...
+        let returnKey: string;
+        Object.keys(obj).forEach((key) => {
+            if (obj[key].includes(searchValue)) {
+                returnKey = key;
+            }
+        });
+        return returnKey;
+    }
+    return searchValue;
+}
```
Comment (krzysztof-causalens) [inline packages/ui-causal-graph-editor/src/shared/rendering/utils.tsx:98]: That's a fairly inefficient implementation of the search, as you're effectively searching for it twice. Can you think of a more efficient implementation? Try to reduce the number of iterations over the lists/objects, and make sure to stop the search whenever an item is found.

**Review by krzysztof-causalens — CHANGES_REQUESTED** [review]
I've had a look through the code so far - will try to find some time to play with it locally soon as well. Code looks good in general, great job! I found some minor style/performance issues below

File: packages/ui-causal-graph-editor/src/graph-viewer/causal-graph-editor.tsx [hunk 2]
```diff
@@ -496,6 +522,25 @@ function CausalGraphEditor({ requireFocusToZoom = true, ...props }: CausalGraphE
         setTooltipContent(edgeTooltipContent);
     });
 
+    useEngineEvent('groupMouseover', (event, groupId) => {
+        // const nodeAttributes = state.graph.getNodeAttributes(nodeId);
```
[the remark this answers is not in the document]
Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/src/graph-viewer/causal-graph-editor.tsx:526]: oops removed

File: packages/ui-causal-graph-editor/src/graph-viewer/causal-graph-editor.tsx [hunk 3]
```diff
@@ -705,13 +750,36 @@ function CausalGraphEditor({ requireFocusToZoom = true, ...props }: CausalGraphE
...
+                                        />
+                                    )}
+                                    {layoutHasGroup && !showCollapseAll && (
+                                        <ExpandAllButton
+                                            onExpandAll={() => {
+                                                setShowCollapseAll(true);
+                                                expandGroups();
+                                            }}
+                                        />
+                                    )}
```
[the remark this answers is not in the document]
Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/src/graph-viewer/causal-graph-editor.tsx:782]: good point fixed

File: packages/ui-causal-graph-editor/src/graph-viewer/custom-layout.stories.tsx
[the remark this answers is not in the document]
Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/src/graph-viewer/custom-layout.stories.tsx:1]: Good point, moved

File: packages/ui-causal-graph-editor/src/shared/rendering/edge/definitions.tsx [hunk 4]
```diff
@@ -39,6 +39,8 @@ export interface PixiEdgeStyle {
     state: EdgeState;
     /** Strength definition to be used if provided */
     strength?: EdgeStrengthDefinition;
+    /** The number of collapsed edges at either end of the edge */
+    collapsedEdges?: number;
```
[the remark this answers is not in the document]
Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/src/shared/rendering/edge/definitions.tsx:43]: Good point, agree

**Comment by claudiawestby** [comment]
getting weird behavior when all the nodes within the same group (towards end of video)
https://github.com/causalens/dara-ui/assets/98322280/eccecb4e-ca07-4850-ac64-82126ccc52ac

**Comment by patricia-causalens** [comment]
> getting weird behavior when all the nodes within the same group (towards end of video) https://github.com/causalens/dara-ui/assets/98322280/eccecb4e-ca07-4850-ac64-82126ccc52ac

Hi @claudiawestby , thanks for the feedback, I have addressed the issue for now to not allow you to remove the last node in a group and show a warning instead (we can change this behaviour in the future, it was easier to implement for now). Could you have another look and let me know if it seems ok?

**Comment by patricia-causalens** [comment]
Merging since will have another PR soon for d3 so can deal with any further issues that come up in that one

**Pull request merged**